java/vector/src/main/java/org/apache/arrow/vector/ipc/ArrowWriter.java

```diff
@@ -207,4 +217,34 @@ public void close() {
       throw new RuntimeException(e);
     }
   }
+
+  /**
+   * Abstract class to configure builder pattern for implementing Arrow writers for IPC over a WriteChannel.
+   * @param <T> Recursive type parameter that allow method chaining tp work properly in subclasses
+   */
+  abstract static class Builder<T extends Builder<T>> {
+    private final VectorSchemaRoot root;
+    private final WriteChannel out;
+    private CompressionCodec codec = NoCompressionCodec.INSTANCE;
+
+    abstract ArrowWriter build();
+
+    protected abstract T self();
+
+    public Builder(VectorSchemaRoot root, WritableByteChannel out) {
+      this.root = root;
+      this.out = new WriteChannel(out);
+    }
+
+    public T setCodec(CompressionCodec codec) {
+      this.codec = codec;
+      return self();
+    }
+  }
+
+  protected ArrowWriter(Builder<?> builder) {
```

Comment:
Ditto here: I'd rather remove this. `build` is abstract so IMO there is no benefit to be had with a common constructor here.